examples/darwin-framework-tool/commands/common/ControllerStorage.mm

```diff
@@ -0,0 +1,159 @@
+
+    __auto_type * controllerKey = [self _keyToControllerKey:key];
+    self.storage[controllerKey] = data;
+    return YES;
+}
+
+- (BOOL)controller:(MTRDeviceController *)controller
+    removeValueForKey:(NSString *)key
+        securityLevel:(MTRStorageSecurityLevel)securityLevel
+          sharingType:(MTRStorageSharingType)sharingType
+{
+#ifdef LOG_DEBUG_CONTROLLER_STORAGE
+    ChipLogError(chipTool, "Controller(%@) Storage - Remove value for %@", controller, key);
+#endif // LOG_DEBUG_CONTROLLER_STORAGE
+
+    __auto_type * controllerKey = [self _keyToControllerKey:key];
+    self.storage[controllerKey] = nil;
+    return YES;
+}
+
+- (NSDictionary<NSString *, id<NSSecureCoding>> *)valuesForController:(MTRDeviceController *)controller securityLevel:(MTRStorageSecurityLevel)securityLevel sharingType:(MTRStorageSharingType)sharingType
+{
+#ifdef LOG_DEBUG_CONTROLLER_STORAGE
+    ChipLogError(chipTool, "Controller(%@) Storage - Get all values", controller);
+#endif // LOG_DEBUG_CONTROLLER_STORAGE
+
+    NSMutableDictionary * valuesToReturn = [NSMutableDictionary dictionary];
+    for (NSString * controllerKey in self.storage) {
+        if (![self _isControllerKey:controllerKey]) {
+            continue;
+        }
+        __auto_type * key = [self _controllerKeyToKey:controllerKey];
+        valuesToReturn[key] = [self controller:controller valueForKey:key securityLevel:securityLevel sharingType:sharingType];
+    }
+
+    if (!valuesToReturn.count) {
+        return nil;
+    }
+
+    return valuesToReturn;
+}
+
+- (BOOL)controller:(MTRDeviceController *)controller storeValues:(NSDictionary<NSString *, id<NSSecureCoding>> *)values securityLevel:(MTRStorageSecurityLevel)securityLevel sharingType:(MTRStorageSharingType)sharingType
+{
+#ifdef LOG_DEBUG_CONTROLLER_STORAGE
+    ChipLogError(chipTool, "Controller(%@) Storage - store values", controller);
+#endif // LOG_DEBUG_CONTROLLER_STORAGE
+
+    for (NSString * key in values) {
+        [self controller:controller storeValue:values[key] forKey:key securityLevel:securityLevel sharingType:sharingType];
+    }
+
+    return YES;
+}
+
+- (NSData *)valueForKey:(NSString *)key
+{
+    __auto_type * controllerKey = [self _keyToControllerKey:key];
+    return self.storage[controllerKey];
+}
+
+- (void)storeValue:(NSData *)value forKey:key
+{
+    __auto_type * controllerKey = [self _keyToControllerKey:key];
+    self.storage[controllerKey] = value;
+}
+
+- (NSString *)_keyToControllerKey:(NSString *)key
+{
+    return [NSString stringWithFormat:@"%@/%@", _controllerID, key];
+}
+
+- (NSString *)_controllerKeyToKey:(NSString *)controllerKey
+{
+    __auto_type * controllerPrefix = [NSString stringWithFormat:@"%@/", [_controllerID UUIDString]];
+    return [controllerKey stringByReplacingOccurrencesOfString:controllerPrefix withString:@""];
+}
+
+- (BOOL)_isControllerKey:(NSString *)controllerKey
+{
+    return [controllerKey hasPrefix:[_controllerID UUIDString]];
```

Comment:
I do wonder whether we should just have a keyScopingPrefix property that has the `[NSString stringWithFormat:@"%@/", [_controllerID UUIDString]]` bit.

Reply:
> I do wonder whether we should just have a keyScopingPrefix property that has the `[NSString stringWithFormat:@"%@/", [_controllerID UUIDString]]` bit.

Great suggestion! I’ve updated the PR to include a `keyScopingPrefix` property with `[NSString stringWithFormat:@"%@/", [_controllerID UUIDString]]` as you suggested.